src/Miki.Framework/Extension/MiddlewareExtensions.cs

```diff
@@ -0,0 +1,152 @@
+﻿﻿using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Miki.Framework.Hosting;
+
+namespace Miki.Framework
+{
+    using Predicate = Func<IContext, bool>;
+    using PredicateAsync = Func<IContext, ValueTask<bool>>;
+    
+    public static class MiddlewareExtensions
+    {
+        /// <summary>
+        /// Adds a middleware delegate defined in-line to the application's request pipeline.
+        /// </summary>
+        /// <param name="app">The <see cref="IBotApplicationBuilder"/> instance.</param>
+        /// <param name="middleware">A function that handles the request or calls the given next function.</param>
+        /// <returns>The <see cref="IBotApplicationBuilder"/> instance.</returns>
+        public static IBotApplicationBuilder Use(this IBotApplicationBuilder app, Func<IContext, Func<ValueTask>, ValueTask> middleware)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+            
+            if (middleware == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+            
+            return app.Use(next =>
+            {
+                return context =>
+                {
+                    return middleware(context, () => next(context));
+                };
+            });
+        }
+        
+        /// <summary>
+        /// Adds a terminal middleware delegate to the application's request pipeline.
+        /// </summary>
+        /// <param name="app">The <see cref="IBotApplicationBuilder"/> instance.</param>
+        /// <param name="handler">A delegate that handles the request.</param>
+        public static void Run(this IBotApplicationBuilder app, MessageDelegate handler)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+
+            app.Use(_ => handler);
```

Comment:
nani?

Reply:
The extension method `.Run` is being used at the end of the pipeline, so you don't need the next middleware. This is the same as ASP.NET Core.